src/github.com/getlantern/flashlight/main/settings.go

```diff
@@ -73,91 +72,114 @@ func Load(version, revisionDate, buildDate string) {
-		defer settings.Unlock()
-		return write(baseSettings)
+		s.Lock()
+		defer s.Unlock()
+		return write(s)
 	}
 	service, err = ui.Register(messageType, nil, helloFn)
 	return err
 }
 
-func read() {
-	log.Tracef("Reading settings messages!!")
+func (s *Settings) read() {
+	log.Debugf("Reading settings messages!!")
 	for message := range service.In {
-		log.Tracef("Read settings message!! %q", message)
+		log.Debugf("Read settings message!! %v", message)
 		msg := (message).(map[string]interface{})
 
 		if autoReport, ok := msg["autoReport"].(bool); ok {
-			SetAutoReport(autoReport)
+			s.SetAutoReport(autoReport)
 		} else if proxyAll, ok := msg["proxyAll"].(bool); ok {
-			SetProxyAll(proxyAll)
+			s.SetProxyAll(proxyAll)
 		} else if autoLaunch, ok := msg["autoLaunch"].(bool); ok {
-			SetAutoLaunch(autoLaunch)
+			s.SetAutoLaunch(autoLaunch)
+		} else if systemProxy, ok := msg["systemProxy"].(bool); ok {
+			log.Debugf("Setting system proxy")
+			s.SetSystemProxy(systemProxy)
 		}
 	}
 }
 
-// Saves settings to disk.
-func Save() {
+// Save saves settings to disk.
+func (s *Settings) Save() {
 	log.Debug("Saving settings")
-	settings.Lock()
-	defer settings.Unlock()
-	if bytes, err := yaml.Marshal(settings); err != nil {
+	s.Lock()
+	defer s.Unlock()
+	if bytes, err := yaml.Marshal(s); err != nil {
 		log.Errorf("Could not create yaml from settings %v", err)
 	} else if err := ioutil.WriteFile(path, bytes, 0644); err != nil {
 		log.Errorf("Could not write settings file %v", err)
 	} else {
-		log.Debugf("Saved settings to %s", path)
+		log.Debugf("Saved settings to %s with contents %v", path, string(bytes))
+	}
+}
+
+// GetProxyAll returns whether or not to proxy all traffic.
+func (s *Settings) GetProxyAll() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.ProxyAll
+}
+
+// SetProxyAll sets whether or not to proxy all traffic.
+func (s *Settings) SetProxyAll(proxyAll bool) {
+	s.Lock()
+	defer s.Unlock()
+	s.ProxyAll = proxyAll
+}
+
+// IsAutoReport returns whether or not to auto-report debugging and analytics data.
+func (s *Settings) IsAutoReport() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.AutoReport
+}
+
+// SetAutoReport sets whether or not to auto-report debugging and analytics data.
+func (s *Settings) SetAutoReport(auto bool) {
+	s.Lock()
+	defer s.Unlock()
+	s.AutoReport = auto
+}
+
+// SetAutoLaunch sets whether or not to auto-launch Lantern on system startup.
+func (s *Settings) SetAutoLaunch(auto bool) {
+	s.Lock()
+	defer s.Unlock()
+	s.AutoLaunch = auto
+	go launcher.CreateLaunchFile(auto)
+}
+
+// GetSystemProxy returns whether or not to set system proxy when lantern starts
+func (s *Settings) GetSystemProxy() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.SystemProxy
+}
+
+// SetSystemProxy sets whether or not to set system proxy when lantern starts
+func (s *Settings) SetSystemProxy(enable bool) {
+	s.Lock()
+	defer s.Unlock()
+	s.SystemProxy = enable
+	if enable {
+		go pacOn()
+	} else {
+		go pacOff()
 	}
```

Comment:
@myleshorton It's very unlikely to happen in practice, but if this function is called in rapid succession, there's no guarantee that these goroutines will run in the right order.  Better would be to set up a chan bool and a single goroutine that toggles based on what it gets from that channel.